src/Hamiltonians/HubbardRealSpace.jl

```diff
@@ -183,126 +318,187 @@ number of sites `M` inferred from the number of modes in `address`.
+  `C × C`. `w[i, j]` corresponds to the interaction between the `i`-th and `j`-th component.
+   `w[i, i]` corresponds to the interaction of a component with itself.
 * `v`: the trap potential strengths. Must be a matrix of size `C × D`. `v[i,j]` is
   the strength of the trap for component `i` in the `j`th dimension.
 """
 struct HubbardRealSpace{
+    TT,
     C, # components
     A<:AbstractFockAddress,
     G<:CubicGrid,
     D, # dimension
-    # The following need to be type params.
-    T<:SVector{C,Float64},
+    # The following need to be type params because SMatrix also includes the length.
+    T<:SMatrix{C,D,TT},
     U<:Union{SMatrix{C,C,Float64},Nothing},
+    W<:Union{SMatrix{C,C,Float64},Nothing},
     V<:Union{SMatrix{C,D,Float64},Nothing},
-    P<:Union{Matrix{Float64},Nothing}
-} <: AbstractHamiltonian{Float64}
+    P<:Union{Matrix{Float64},Nothing},
+} <: AbstractHamiltonian{TT}
     address::A
     t::T # hopping strengths
     u::U # interactions
+    w::W # nearest neighbour interactions
     v::V # trap strengths
     potential::P # potential energy of each component at each lattice site
     geometry::G
 end
 
 function HubbardRealSpace(
     address::AbstractFockAddress;
-    geometry::CubicGrid=PeriodicBoundaries((num_modes(address),)),
-    t=ones(num_components(address)),
-    u=ones(num_components(address), num_components(address)),
-    v=zeros(num_components(address), num_dimensions(geometry))
+    geometry::CubicGrid=CubicGrid(num_modes(address)),
+    t=1.0,
+    u=1.0,
+    w=0.0,
+    v=0.0,
 )
     C = num_components(address)
     D = num_dimensions(geometry)
     S = size(geometry)
+    TT = float(eltype(t))
 
     # Sanity checks
     if prod(size(geometry)) ≠ num_modes(address)
         throw(ArgumentError("`geometry` does not have the correct number of sites"))
-    elseif length(u) ≠ 1 && !issymmetric(u)
-        throw(ArgumentError("`u` must be symmetric"))
-    elseif length(u) ≠ C * C
-        throw(ArgumentError("`u` must be a $C × $C matrix"))
-    elseif size(t) ≠ (C,)
-        throw(ArgumentError("`t` must be a vector of length $C"))
-    elseif length(v) ≠ C * D
-        throw(ArgumentError("`v` must be a $C × $D matrix"))
     elseif !(address isa SingleComponentFockAddress || address isa CompositeFS)
-        throw(ArgumentError(
-            "unsupported address type detected use `CompositeFS` or `<: SingleComponentFockAddress`"
-        ))
+        throw(
+            ArgumentError(
+                "unsupported address type detected use `CompositeFS` or `<: SingleComponentFockAddress`",
+            ),
+        )
     end
-    warn_fermi_interaction(address, u)
+    t_mat = _t_or_v_to_matrix(:t, t, C, D; zero_is_nothing=false)
+    u_mat = _u_or_w_to_matrix(:u, u, C)
+    w_mat = _u_or_w_to_matrix(:w, w, C)
+    v_mat = _t_or_v_to_matrix(:v, v, C, D)
 
-    t_vec = SVector{C,Float64}(t)
-    u_mat = iszero(u) ? nothing : SMatrix{C,C,Float64}(u)
+    warn_fermi_interaction(address, u_mat)
```

Comment:
As an alternative you could do the check like this and not modify `warn_fermi_interction`.
```suggestion
    u ≢ 1.0 && warn_fermi_interaction(address, u_mat)
```